sklearn/linear_model/logistic.py

```diff
@@ -1557,32 +1554,34 @@ def fit(self, X, y, sample_weight=None):
 
         X, y = check_X_y(X, y, accept_sparse='csr', dtype=np.float64,
                          order="C")
-
+        check_classification_targets(y)
+        self.classes_ = np.unique(y)
         if self.solver == 'sag':
             max_squared_sum = row_norms(X, squared=True).max()
         else:
             max_squared_sum = None
 
-        check_classification_targets(y)
-
-        if y.ndim == 2 and y.shape[1] == 1:
-            warnings.warn(
-                "A column-vector y was passed when a 1d array was"
-                " expected. Please change the shape of y to "
-                "(n_samples, ), for example using ravel().",
-                DataConversionWarning)
-            y = np.ravel(y)
-
-        check_consistent_length(X, y)
+        if y.dtype.kind in ('S', 'U'):
+            # Encode for string labels
+            self._label_encoder = LabelEncoder().fit(y)
+            y = self._label_encoder.transform(y)
+            self.classes_enc_ = np.arange(len(self._label_encoder.classes_))
+            if isinstance(self.class_weight, dict):
+                old_keys = self.class_weight.keys()
+                new_keys = self._label_encoder.transform(old_keys)
+                # Don't modify the original class_weight dict.
+                self.class_weight = dict()
+                for new_key, old_key in zip(new_keys, old_keys):
+                    self.class_weight[new_key] = self.class_weight[old_key]
+        else:
+            self.classes_enc_ = self.classes_
```

Comment:
I'm encoding for all cases now like you suggested. And there is a test for this usecase.